### PR TITLE
arch: arc: fix the bug in prologue of sys call handling

### PR DESCRIPTION
**Pull request by vonhust**
* the old codes may not save the caller saved regs correctly, e.g. r7- r12.
  Because the sys call entry is called in the form of static inline function.
  The compiler optimizations may not save all the caller saved regs.

* new codes use the irq stack frame as the sys call frame and gurantee all the
  called saved regs are pushed and popped correctly.

* the side effect of new codes are more stack operations and a little overhead.s

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

### CODE REVIEW

File: arch/arc/core/fault_s.S [hunk 1]
```diff
@@ -205,21 +205,23 @@ SECTION_SUBSEC_FUNC(TEXT,__fault,__ev_trap)
 	mov_s r6, K_SYSCALL_BAD
 
 valid_syscall_id:
+	_create_irq_stack_frame
+
 #ifdef CONFIG_ARC_SECURE_FIRMWARE
-	lr ilink, [_ARC_V2_ERSEC_STAT]
-	push ilink
+	/* ERSEC_STAT is IOW/RAZ in normal mode */
+	lr r0,[_ARC_V2_ERSEC_STAT]
```
Comment (abrodkin) [inline arch/arc/core/fault_s.S]: That's obviously a nitpicking but could you please keep spaces in between "," and the next symbol - it makes code much more readable. BTW I'm wondering if checkpatch checks assembly code or not? Have you run checkpatch on your changes?

File: arch/arc/core/fault_s.S [hunk 2]
```diff
@@ -205,21 +205,23 @@ SECTION_SUBSEC_FUNC(TEXT,__fault,__ev_trap)
 	mov_s r6, K_SYSCALL_BAD
 
 valid_syscall_id:
+	_create_irq_stack_frame
+
```
Comment (abrodkin) [inline arch/arc/core/fault_s.S]: I would add here a comment explaining that we save `r0` in `_create_irq_stack_frame()` and so now we may safely use it for other things with no fear or losing its initial contents.

File: arch/arc/core/fault_s.S [hunk 3]
```diff
@@ -236,7 +238,7 @@ _do_non_syscall_trap:
 	/* save caller saved registers */
 	_create_irq_stack_frame
 
-#ifdef CONFIG_ARC_HAS_SECURE
+#ifdef CONFIG_ARC_SECURE_FIRMWARE
```
Comment (abrodkin) [inline arch/arc/core/fault_s.S]: I'd suggest to leave this change to a separate commit as it has nothing to do with that particular fix.

**Review by abrodkin — CHANGES_REQUESTED** [review]
Please add spaces after "," and move `CONFIG_ARC_SECURE` -> `CONFIG_ARC_SECURE_FIRMWARE` to a separate commit.

File: arch/arc/core/fault_s.S (hunk 1 above)
Comment (vonhust) [reply arch/arc/core/fault_s.S]: Fixed.

I ran checkpatch, it seems that it doesn't check assembly code

File: arch/arc/core/fault_s.S (hunk 2 above)
Comment (vonhust) [reply arch/arc/core/fault_s.S]: done

File: arch/arc/core/fault_s.S (hunk 3 above)
Comment (vonhust) [reply arch/arc/core/fault_s.S]: agreed.

File: arch/arc/core/fault_s.S (hunk 2 above)
Comment (abrodkin) [reply arch/arc/core/fault_s.S]: Well it's obvious to you and me that caller regs are saved in IRQ stack frame but new reader might be confused.
That said maybe explicitly mention that `r0` is saved in `_create_irq_stack_frame()`?

Also looks like comments indentation in other places of the same source file is different (+1 tab) compared to ones above and below this place.

**Review by abrodkin — APPROVED** [review]
LGTM now, good to merge.

**Review by ruuddw — APPROVED** [review]
Would other architectures have a similar problem? Or don't they have a similar optimization to avoid saving  caller saved regs twice?

**Comment by andrewboie** [comment]
This needs to be backported to the 1.14 branch, this is a very serious bug.
